main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -501,6 +493,42 @@ object Inliner {
     }
   }
 
+  /**
+   * Performs boolean folding on a given binary expression with the logic:
+   * Only fold if the expression removed is pure, and
+   * For Or-expressions,
+   * fold into true if either the left or right expression is true.
+   * if either the left or right expression is false, fold into the other expression.
+   * For And-expressions,
+   * fold into false, if either the left or right expression is false.
+   * if either the left or right expression is true, fold into the other expression.
+   */
+  private def binaryFold(exp0: LiftedAst.Expression.Binary): LiftedAst.Expression = {
+    (exp0.sop, exp0.exp1, exp0.exp2) match {
+      case (SemanticOperator.BoolOp.And, LiftedAst.Expression.True(_), _) => exp0.exp2
+      case (SemanticOperator.BoolOp.And, _, LiftedAst.Expression.True(_)) => exp0.exp1
+      case (SemanticOperator.BoolOp.And, LiftedAst.Expression.False(_), _) => LiftedAst.Expression.False(exp0.loc)
+      case (SemanticOperator.BoolOp.And, _, LiftedAst.Expression.False(_)) if exp0.exp1.purity == Pure => LiftedAst.Expression.False(exp0.loc)
+      case (SemanticOperator.BoolOp.Or, LiftedAst.Expression.False(_), _) => exp0.exp2
+      case (SemanticOperator.BoolOp.Or, _, LiftedAst.Expression.False(_)) => exp0.exp1
+      case (SemanticOperator.BoolOp.Or, LiftedAst.Expression.True(_), _) => LiftedAst.Expression.True(exp0.loc)
+      case (SemanticOperator.BoolOp.Or, _, LiftedAst.Expression.True(_)) if exp0.exp1.purity == Pure => LiftedAst.Expression.True(exp0.loc)
+      case _ => exp0
+    }
+  }
+
+  /**
+   * Performs boolean folding on a given unary expression with the logic:
+   * Folds not-true => false and not-false => true.
+   */
+  private def unaryFold(exp0: LiftedAst.Expression.Unary): LiftedAst.Expression = {
```

Comment:
This should occur before binaryFold to keep the right order.

Reply:
I think both of these helper functions should occur further down in the file, towards the bottom.